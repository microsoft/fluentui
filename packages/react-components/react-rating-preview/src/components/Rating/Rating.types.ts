import * as React from 'react';
import type { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';

export type RatingSlots = {
  root: NonNullable<Slot<'div'>>;
  ratingLabel?: NonNullable<Slot<'label'>>;
  ratingCountLabel?: NonNullable<Slot<'label'>>;
};

/**
 * Rating Props
 */
export type RatingProps = ComponentProps<RatingSlots> & {
  /**
   * Controls the appearance of the Rating.
   * @default neutral
   */
  color?: 'brand' | 'marigold' | 'neutral';
  /**
   * Default value of the Rating
   */
  defaultValue?: number;
  /**
   * The icon to display when the rating value is greater than or equal to the item's value.
   */
  iconFilled?: React.ReactElement;
  /**
   * The icon to display when the rating value is less than the item's value.
   */
  iconOutline?: React.ReactElement;
  /**
   * The max value of the rating. This controls the number of rating items displayed.
   * Must be a whole number greater than 1.
   * @default 5
   */
  max?: number;
  /**
   * The mode of the rating.
   * @default 'interactive'
   */
  mode?: 'interactive' | 'read-only' | 'read-only-compact';
  /**
   * Name for the Radio inputs. If not provided, one will be automatically generated
   */
  name?: string;
  /**
   * Callback when the rating value is changed by the user.
   */
  onChange?: (ev: React.SyntheticEvent | Event, data: RatingOnChangeData) => void;
  /**
   * Sets the precision to allow half-filled shapes in Rating
   * @default 1
   */
  step?: 0.5 | 1;
  /**
   * Sets the size of the Rating items.
   * @default medium
   */
  size?: 'small' | 'medium' | 'large' | 'extra-large';
  /**
   * The value of the rating
   */
  value?: number;
};

/**
 * Data for the onChange event for Rating.
 */
export type RatingOnChangeData = {
  /**
   * The new value of the rating.
   */
  value?: number;
};

/**
 * State used in rendering Rating
 */
export type RatingState = ComponentState<RatingSlots> &
  Required<Pick<RatingProps, 'color' | 'iconFilled' | 'iconOutline' | 'mode' | 'name' | 'step' | 'size' | 'value'>> & {
    hoveredValue?: number | undefined;
  };

export type RatingContextValue = Pick<
  RatingState,
  'color' | 'iconFilled' | 'iconOutline' | 'mode' | 'name' | 'step' | 'size' | 'value' | 'hoveredValue'
>;

export type RatingContextValues = {
  rating: RatingContextValue;
};
