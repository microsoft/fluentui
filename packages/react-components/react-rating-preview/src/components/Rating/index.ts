export * from './Rating';
export * from './Rating.types';
export * from './renderRating';
export * from './useRating';
export * from './useRatingStyles.styles';
