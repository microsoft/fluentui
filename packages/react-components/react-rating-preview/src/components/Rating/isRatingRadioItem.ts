import { isHTMLElement } from '@fluentui/react-utilities';

export const isRatingRadioItem = (name: string, target: EventTarget): target is HTMLInputElement =>
  isHTMLElement(target, { constructorName: 'HTMLInputElement' }) && target.type === 'radio' && target.name === name;
