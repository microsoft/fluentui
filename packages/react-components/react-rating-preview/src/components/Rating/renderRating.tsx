/** @jsxRuntime automatic */
/** @jsxImportSource @fluentui/react-jsx-runtime */

import { assertSlots } from '@fluentui/react-utilities';
import type { RatingState, RatingSlots, RatingContextValues } from './Rating.types';
import { RatingProvider } from '../../contexts/RatingContext';

/**
 * Render the final JSX of Rating
 */
export const renderRating_unstable = (state: RatingState, contextValues: RatingContextValues) => {
  assertSlots<RatingSlots>(state);

  return (
    <RatingProvider value={contextValues.rating}>
      <state.root>
        {state.root.children}
        {state.ratingLabel && <state.ratingLabel />}
        {state.ratingCountLabel && <state.ratingCountLabel />}
      </state.root>
    </RatingProvider>
  );
};
