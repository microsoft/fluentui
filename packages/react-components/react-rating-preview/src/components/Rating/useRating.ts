import * as React from 'react';
import { getIntrinsicElementProps, slot } from '@fluentui/react-utilities';
import type { RatingProps, RatingState } from './Rating.types';

/**
 * Create the state required to render Rating.
 *
 * The returned state can be modified with hooks such as useRatingStyles_unstable,
 * before being passed to renderRating_unstable.
 *
 * @param props - props from this instance of Rating
 * @param ref - reference to root HTMLElement of Rating
 */
export const useRating_unstable = (props: RatingProps, ref: React.Ref<HTMLDivElement>): RatingState => {
  return {
    // TODO add appropriate props/defaults
    components: {
      // TODO add each slot's element type or component
      root: 'div',
    },
    // TODO add appropriate slots, for example:
    // mySlot: resolveShorthand(props.mySlot),
    root: slot.always(
      getIntrinsicElementProps('div', {
        ref,
        ...props,
      }),
      { elementType: 'div' },
    ),
  };
};
