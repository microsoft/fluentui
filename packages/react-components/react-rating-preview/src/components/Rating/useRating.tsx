import * as React from 'react';
import {
  getIntrinsicElementProps,
  isHTMLElement,
  mergeCallbacks,
  slot,
  useControllableState,
  useId,
} from '@fluentui/react-utilities';
import type { RatingProps, RatingState } from './Rating.types';
import { RatingItem } from '../../RatingItem';
import { StarFilled, StarRegular } from '@fluentui/react-icons';

/**
 * Create the state required to render Rating.
 *
 * The returned state can be modified with hooks such as useRatingStyles_unstable,
 * before being passed to renderRating_unstable.
 *
 * @param props - props from this instance of Rating
 * @param ref - reference to root HTMLElement of Rating
 */
export const useRating_unstable = (props: RatingProps, ref: React.Ref<HTMLDivElement>): RatingState => {
  const generatedName = useId('rating-');
  const {
    color = 'neutral',
    iconFilled = <StarFilled />,
    iconOutline = <StarRegular />,
    max = 5,
    name = generatedName,
    onChange,
    step = 1,
    size = 'extra-large',
  } = props;

  const [value, setValue] = useControllableState({
    state: props.value,
    defaultState: props.defaultValue,
    initialState: 0,
  });

  const isRatingRadioItem = (target: EventTarget): target is HTMLInputElement =>
    isHTMLElement(target, { constructorName: 'HTMLInputElement' }) && target.type === 'radio' && target.name === name;

  const [hoveredValue, setHoveredValue] = React.useState<number | undefined>(undefined);

  // Generate the child RatingItems and memoize them to prevent unnecessary re-rendering
  const rootChildren = React.useMemo(() => {
    return Array.from(Array(max), (_, i) => <RatingItem value={i + 1} key={i + 1} />);
  }, [max]);

  const state: RatingState = {
    color,
    iconFilled,
    iconOutline,
    name,
    step,
    size,
    value,
    hoveredValue,
    components: {
      root: 'div',
    },
    root: slot.always(
      getIntrinsicElementProps(
        'div',
        {
          ref,
          children: rootChildren,
          role: 'radiogroup',
          ...props,
        },
        ['onChange'],
      ),
      { elementType: 'div' },
    ),
  };

  state.root.onChange = ev => {
    if (isRatingRadioItem(ev.target)) {
      const newValue = parseFloat(ev.target.value);
      if (!isNaN(newValue)) {
        setValue(newValue);
        onChange?.(ev, { type: 'change', event: ev, value: newValue });
      }
    }
  };
  state.root.onMouseOver = mergeCallbacks(props.onMouseOver, ev => {
    if (isRatingRadioItem(ev.target)) {
      const newValue = parseFloat(ev.target.value);
      if (!isNaN(newValue)) {
        setHoveredValue(newValue);
      }
    }
  });
  state.root.onMouseLeave = mergeCallbacks(props.onMouseLeave, ev => {
    setHoveredValue(undefined);
  });

  return state;
};
