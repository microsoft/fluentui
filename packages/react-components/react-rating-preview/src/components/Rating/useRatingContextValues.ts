import * as React from 'react';
import { RatingContextValues, RatingState } from './Rating.types';
import { RatingItemContextValue } from '../RatingItem/RatingItem.types';

export const useRatingContextValues = (ratingState: RatingState): RatingContextValues => {
  const { color, hoveredValue, iconFilled, iconOutline, itemLabel, name, step, size, value } = ratingState;

  const ratingItem = React.useMemo<RatingItemContextValue>(
    () => ({
      color,
      hoveredValue,
      iconFilled,
      iconOutline,
      interactive: true,
      itemLabel,
      name,
      step,
      size,
      value,
    }),
    [color, hoveredValue, iconFilled, iconOutline, itemLabel, name, step, size, value],
  );

  return { ratingItem };
};
