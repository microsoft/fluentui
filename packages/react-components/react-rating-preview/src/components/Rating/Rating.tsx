import * as React from 'react';
import type { ForwardRefComponent } from '@fluentui/react-utilities';
import { useRating_unstable } from './useRating';
import { renderRating_unstable } from './renderRating';
import { useRatingStyles_unstable } from './useRatingStyles.styles';
import type { RatingProps } from './Rating.types';
import { useRatingContextValues } from '../../contexts/useRatingContextValues';

/**
 * Rating component - TODO: add more docs
 */
export const Rating: ForwardRefComponent<RatingProps> = React.forwardRef((props, ref) => {
  const state = useRating_unstable(props, ref);
  const contextValues = useRatingContextValues(state);

  useRatingStyles_unstable(state);
  return renderRating_unstable(state, contextValues);
});

Rating.displayName = 'Rating';
