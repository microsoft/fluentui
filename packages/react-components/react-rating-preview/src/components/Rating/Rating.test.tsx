import * as React from 'react';
import { render } from '@testing-library/react';
import { isConformant } from '../../testing/isConformant';
import { Rating } from './Rating';

describe('Rating', () => {
  isConformant({
    Component: Rating,
    displayName: 'Rating',
    // TODO re-enable consistent-callback-args once https://github.com/microsoft/fluentui/pull/30376 is merged
    disabledTests: ['consistent-callback-args'],
  });
  it('respects a default value', () => {
    const { getAllByRole } = render(<Rating defaultValue={3} />);
    const checkedItems = getAllByRole('radio').filter(item => (item as HTMLInputElement).checked);
    expect(checkedItems[0].getAttribute('value')).toBe('3');
    expect(checkedItems.length).toEqual(1);
  });
  it('only sets the selected rating item to checked', () => {
    const { getAllByRole } = render(<Rating value={3} />);
    const checkedItems = getAllByRole('radio').filter(item => (item as HTMLInputElement).checked);
    expect(checkedItems[0].getAttribute('value')).toBe('3');
    expect(checkedItems.length).toEqual(1);
  });
  it('renders the correct number of items', () => {
    const { getAllByRole } = render(<Rating max={10} />);
    const items = getAllByRole('radio');
    expect(items.length).toEqual(10);
  });
  it('calle onChange when a rating is clicked', () => {
    const onChange = jest.fn();
    const { getAllByRole } = render(<Rating onChange={onChange} />);
    const items = getAllByRole('radio');
    items[0].click();
    expect(onChange).toHaveBeenCalledTimes(1);
  });
  it('does not call onChange when a rating is clicked and the value is the same', () => {
    const onChange = jest.fn();
    const { getAllByRole } = render(<Rating value={3} onChange={onChange} />);
    const items = getAllByRole('radio');
    items[2].click();
    expect(onChange).toHaveBeenCalledTimes(0);
  });
  it('calls onChange with the correct value when a rating is clicked', () => {
    const onChange = jest.fn();
    const { getAllByRole } = render(<Rating onChange={onChange} />);
    const items = getAllByRole('radio');
    items[3].click();
    items[2].click();
    items[1].click();
    expect(onChange).toHaveBeenCalledTimes(3);
    expect(onChange.mock.calls[0][1].value).toBe(4);
    expect(onChange.mock.calls[1][1].value).toBe(3);
    expect(onChange.mock.calls[2][1].value).toBe(2);
  });
});
