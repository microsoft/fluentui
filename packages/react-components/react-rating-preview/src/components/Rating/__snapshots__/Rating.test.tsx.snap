// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rating renders a default state 1`] = `
<div>
  <div
    class="fui-Rating"
  >
    Default RatingItem
  </div>
</div>
`;
